middleware/common/include/common/transcode.h

```diff
@@ -17,266 +17,296 @@
+      namespace utf8
+      {
+         //! @param value String encoded in local default codeset
+         //!
+         //! @return UTF-8-encoded string
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string encode( const std::string& value);
+
+         //! @param value The UTF-8 encoded string
+         //!
+         //! @return String encoded in local default codeset
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string decode( const std::string& value);
+
+         //! @param codeset String-encoding
+         //!
+         //! @return Whether the provided codeset exist in the system
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::Casual on other failures
+         bool exist( std::string_view codeset);
+
+         //! @param value String encoded in provided codeset
+         //! @param codeset String-encoding
+         //!
+         //! @return UTF-8-encoded string
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string encode( const std::string& value, std::string_view codeset);
+
+         //! @param value The UTF-8 encoded string
+         //! @param codeset Encoding for result
+         //!
+         //! @return String encoded in provided codeset
+         //!
+         //! @throw exception::limit::Memory on resource failures
+         //! @throw exception::system::invalid::Argument for bad input
+         //! @throw exception::Casual on other failures
+         std::string decode( const std::string& value, std::string_view codeset);
+
+      } // utf8
+
+      namespace hex
+      {
+         namespace detail
+         {
+            template< typename Input, typename Out>
+            void encode( Input first, Input last, Out out)
             {
-               auto cast = []( auto&& i){ return reinterpret_cast< const char*>( &(*i));}; 
-               detail::encode( out, cast( first), cast( last));
-               return out;
-            }
+               auto hex = []( auto value)
+               {
+                  if( value < 10)
+                     return value + 48;
+                  return value + 87;
+               };
 
-            template< typename R>
-            auto encode( std::ostream& out, R&& range) -> 
-               std::enable_if_t< traits::is::binary::like_v< R>, std::ostream&>
-            {
-               return encode( out, std::begin( range), std::end( range));
+               for( ; first != last; ++first)
```

Comment:
Is this some "standard" formula when encoding? Looks a little like voodoo to me. ;-)

Reply:
It's not that voodoish :)  bitmask the upper 4 bits -> right shift 4 places (bits) (so it becomes the lower 4 bits, hence has a value of 0 to 2 ^ 4) -> transform the value to a hex-friendly char.  Do the same for the lower 4 bits, but we don't need to bit-shirt any thing, since the 4 bits are already in the "right" place.

Reply:
Works as long as the character sets are "ASCII" based. Would not work with EBCDIC. In practice only used in IBM mainframes (as far as I know. May be some other odd systems out there) so I do not think we need to worry too much :-)
EBCDIC has digits starting at 0xF0, and A-Z (and corresponding lower case characters) are split into 3 separate ranges, and lower case 'a' is not 0x61.